hexrd/wppf/WPPF.py

```diff
@@ -434,77 +443,87 @@ def computespectrum(self):
 
             for k, l in self.phases.wavelength.items():
 
+                name = self.phases[p].name
+                lam = l[0].getVal("nm")
                 Ic = self.Icalc[p][k]
 
                 shft_c = np.cos(0.5 * np.radians(self.tth[p][k])) * self.shft
                 trns_c = np.sin(np.radians(self.tth[p][k])) * self.trns
-                tth = self.tth[p][k] + self.zero_error + shft_c + trns_c
+
+                sf_shift = 0.0
+                Xs = np.zeros(Ic.shape)
+                if self.phases[p].sf_alpha is not None:
+                    alpha = getattr(self, f"{p}_sf_alpha")
+                    beta  = getattr(self, f"{p}_twin_beta")
+                    sf_shift = alpha*np.tan(np.radians(self.tth[p][k])) *\
+                        self.sf_hkl_factors[p][k]
+                    Xs = np.degrees(
+                        0.9*(1.5*alpha+beta)*(
+                        self.sf_lfactor[p][k]*lam/self.phases[p].lparms[0]))
+
+                tth = self.tth[p][k] + self.zero_error + \
+                    shft_c + trns_c + sf_shift
 
                 dsp = self.dsp[p][k]
                 hkls = self.hkls[p][k]
                 # n = np.min((tth.shape[0], Ic.shape[0]))
                 shkl = self.phases[p].shkl
-                name = self.phases[p].name
-                eta_n = f"self.{name}_eta_fwhm"
-                eta_fwhm = eval(eta_n)
-                strain_direction_dot_product = 0.0
-                is_in_sublattice = False
+                eta_fwhm = getattr(self, f"{name}_eta_fwhm")
+
+                X = getattr(self, f"{name}_X")
+                Y = getattr(self, f"{name}_Y")
+                P = getattr(self, f"{name}_P")
+                XY = np.array([X, Y])
 
                 if self.peakshape == 0:
                     args = (
                         np.array([self.U, self.V, self.W]),
-                        self.P,
-                        np.array([self.X, self.Y]),
-                        np.array([self.Xe, self.Ye, self.Xs]),
+                        P,
+                        XY,
+                        Xs,
                         shkl,
                         eta_fwhm,
                         self.HL,
                         self.SL,
                         tth,
                         dsp,
                         hkls,
-                        strain_direction_dot_product,
-                        is_in_sublattice,
                         tth_list,
                         Ic,
                         self.xn,
-                        self.wn,
+                        self.wn
```

Comment:
I think comma at the end is good practice. You can keep it in the future.

Reply:
Fixed!